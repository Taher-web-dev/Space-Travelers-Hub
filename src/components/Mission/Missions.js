import { React, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import './mission.css';
import { allowJoinMission, allowLeaveMission } from '../../redux/missions/actions/actions';

const Missions = () => {
  const dispatch = useDispatch();
  const missions = useSelector((state) => state.missions.missions);
  const adjustMarginButton = () => {
    const btns = Array.from(document.querySelectorAll('button'));
    btns.forEach((btn) => {
      const parent = btn.parentNode.parentNode;
      const h = parent.offsetHeight;
      const margin = `${0.4 * parseInt(h, 10)}`;
      const elements = parent.querySelectorAll('button');
      for (let i = 0; i < elements.length; i += 1) {
        const elem = elements[i];
        if (elem.className.includes('mission')) elem.style.margin = `${margin - 6}px 10% `;
        else elem.style.margin = `${margin}px 0`;
      }
    });
  };
  const joinMissionHandler = (id) => {
    dispatch(allowJoinMission(id));
  };
  const leaveMissionHandler = (id) => {
    dispatch(allowLeaveMission(id));
  };
  useEffect(() => adjustMarginButton(), [missions]);
  return (
    <table style={{ borderCollapse: 'collapse', width: '90%', margin: '1.25% 20px 5% 5%' }}>
      <thead>
        <tr>
          <th style={{ width: '11.7%' }}>Mission</th>
          <th style={{ width: '58.8%' }}>Description</th>
          <th style={{ width: '11.7%' }}>Status</th>
          <th> </th>
        </tr>
      </thead>
      <tbody>
        {missions.map((mission) => {
          const {
            id, name, description, reserved,
          } = mission;
          return (
            <tr key={id} style={(reserved) ? { backgroundColor: '#fff' } : { backgroundColor: 'rgb(242, 242, 242)' }}>
              <td className="mission-name">{name}</td>
              <td className="description">{description}</td>
              <td>
                {reserved
            && (
            <button type="submit" className="member active-member"> Active Member</button>)}
                {!reserved
            && (<button type="submit" className="member not-member"> NOT A MEMBER</button>)}
              </td>
              <td>
                {!reserved
              && (<button type="submit" className="mission join-mission" onClick={() => joinMissionHandler(id)}>Join Mission</button>)}
                { reserved
               && (<button type="submit" className="mission leave-mission" onClick={() => leaveMissionHandler(id)}>Leave Mission</button>
               )}
              </td>

            </tr>
          );
        })}
      </tbody>
    </table>
  );
};

export default Missions;
