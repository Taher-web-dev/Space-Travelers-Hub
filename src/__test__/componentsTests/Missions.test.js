import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../redux/configureStore';
import Missions from '../../components/Mission/Missions';

describe('Missions', () => {
  it('renders correctly', () => {
    const tree = render(<Provider store={store}><Missions /></Provider>);
    expect(tree).toMatchSnapshot();
  });
});
