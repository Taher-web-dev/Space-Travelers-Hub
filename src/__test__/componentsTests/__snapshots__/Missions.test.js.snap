// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table
        style="border-collapse: collapse; width: 90%; margin: 1.25% 20px 5% 5%;"
      >
        <thead>
          <tr>
            <th
              style="width: 11.7%;"
            >
              Mission
            </th>
            <th
              style="width: 58.8%;"
            >
              Description
            </th>
            <th
              style="width: 11.7%;"
            >
              Status
            </th>
            <th>
               
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </body>,
  "container": <div>
    <table
      style="border-collapse: collapse; width: 90%; margin: 1.25% 20px 5% 5%;"
    >
      <thead>
        <tr>
          <th
            style="width: 11.7%;"
          >
            Mission
          </th>
          <th
            style="width: 58.8%;"
          >
            Description
          </th>
          <th
            style="width: 11.7%;"
          >
            Status
          </th>
          <th>
             
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
